service/Core/DataFormats/WebPages/HtmlDecoder.cs

```diff
@@ -1,32 +1,55 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 
 using System;
+using System.Collections.Generic;
 using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
 using HtmlAgilityPack;
+using Microsoft.Extensions.Logging;
+using Microsoft.KernelMemory.Diagnostics;
+using Microsoft.KernelMemory.Pipeline;
 
 namespace Microsoft.KernelMemory.DataFormats.WebPages;
 
-public class HtmlDecoder
+public class HtmlDecoder : IContentDecoder
 {
-    public FileContent ExtractContent(string filename)
+    private readonly ILogger<HtmlDecoder> _log;
+
+    public IEnumerable<string> SupportedMimeTypes { get; } = new[] { MimeTypes.Html };
+
+    public HtmlDecoder(ILogger<HtmlDecoder>? log = null)
+    {
+        this._log = log ?? DefaultLogger<HtmlDecoder>.Instance;
+    }
+
+    public Task<FileContent> ExtractContentAsync(string filename, string mimeType, CancellationToken cancellationToken = default)
     {
         using var stream = File.OpenRead(filename);
-        return this.ExtractContent(stream);
+        return this.ExtractContentAsync(Path.GetFileName(filename), stream, mimeType, cancellationToken);
     }
 
-    public FileContent ExtractContent(BinaryData data)
+    public Task<FileContent> ExtractContentAsync(string name, BinaryData data, string mimeType, CancellationToken cancellationToken = default)
     {
         using var stream = data.ToStream();
-        return this.ExtractContent(stream);
+        return this.ExtractContentAsync(name, stream, mimeType, cancellationToken);
     }
 
-    public FileContent ExtractContent(Stream data)
+    public Task<FileContent> ExtractContentAsync(string name, Stream data, string mimeType, CancellationToken cancellationToken = default)
     {
+        this._log.LogDebug("Extracting text from HTML file {0}", name);
+
         var doc = new HtmlDocument();
         doc.Load(data);
 
         var result = new FileContent();
+        if (mimeType == MimeTypes.MarkDown)
```

Comment:
Yes, if you can take care of this change I really appreciate it. I have understand what you mean, but I'm not sure how to implement it in an efficient way ;)

Reply:
cool, I'll wait for your changes to be complete, could you drop me a msg here to let me know, so we don't step on each other :-)

Reply:
Sorry, there was an error in the code. The check for the `MarkDown` Mime type must be done in the [TextDecoder.cs class](https://github.com/marcominerva/kernel-memory/blob/0447f6b788dc6a9f10f08f86e355dd43ccec9dcc/service/Core/DataFormats/Text/TextDecoder.cs#L46-L49), because it support multiple types (`PlainText`, `MarkDown`, and `Json`). I have updated it and removed the corresponding lines from **HtmlDecoder.cs**, that supports only `Html` files.

So, `mimeType` is still a required parameter for Decoders. I think it is a good idea to keep this parameter in any case, so that, if necessary, a decoder that supports multiple Mime types can easily determine the actual type on which it is invoked.

If, however, you think it's better to remove it, I can create a specific `MarkDownDecoder` that handles only this Mime type and then set the `MimeType` property of `FileContent` to the correct value.

Reply:
I see pros and cons:
* if we don't pass it, it will invite developers writing small decoders for a specific type, e.g. making it impossible to write a gigantic decoder that handle all types -- this would be good
* if we pass it, we should validate that it is in the list of supported types

having to make a choice, I would not pass it, and have specific decoders, easier to review and manage long term I think. In fact there's some work to do with JSON files, and having a specific decoder would help

Reply:
fixed